### PR TITLE
Fix style getting resseted on tablet/mobile previews

### PR DESCRIPTION
**Pull request by Soare-Robert-Daniel**
TODO:
 - [x] Editor
 - [x] Frontend

### CODE REVIEW

**Comment by HardeepAsrani** [comment]
You need to make sure this script is only loaded during FSE.

File: inc/css/class-block-frontend.php [hunk 1]
```diff
@@ -349,6 +349,22 @@ public function get_post_css( $post_id = '' ) {
...
+
+				$blocks = parse_blocks( $template->content );
+				if ( ! is_array( $blocks ) || empty( $blocks ) ) {
+					continue;
+				}
+				$template_css .= $this->cycle_through_blocks( $blocks );
+			}
+
+			$css .= $template_css;
+
```
[the remark this answers is not in the document]
Comment (HardeepAsrani) [reply inc/css/class-block-frontend.php:367]: @Soare-Robert-Daniel I'm a little lost here, can you explain with more context what you mean by "get only the template that belongs to a post?"

[the remark this answers is not in the document]
Comment (Soare-Robert-Daniel) [reply inc/css/class-block-frontend.php:367]: You can define templates parts like `header` and `footer` that are a part of a post. We can get `footer` and `header` template part via `[block_template_part](https://developer.wordpress.org/reference/functions/block_template_part/)`. But what about the `site` and the `templates`? If you are on 404 page, we should parse only the templates that belong to that page. 
image
image

[the remark this answers is not in the document]
Comment (HardeepAsrani) [reply inc/css/class-block-frontend.php:367]: @Soare-Robert-Daniel You might find something here: https://developer.wordpress.org/reference/files/wp-includes/block-template-utils.php/

[the remark this answers is not in the document]
Comment (Soare-Robert-Daniel) [reply inc/css/class-block-frontend.php:367]: That is where I found `get_block_templates`. But, I did not see a selection functions that pulled the templates based on the post.

[the remark this answers is not in the document]
Comment (HardeepAsrani) [reply inc/css/class-block-frontend.php:367]: Something like this? https://developer.wordpress.org/reference/functions/_build_template_result_from_post/

[the remark this answers is not in the document]
Comment (Soare-Robert-Daniel) [reply inc/css/class-block-frontend.php:367]: It's ok to use it? `This function’s access is marked private. This means it is not intended for use by plugin or theme developers, only in other core functions. It is listed here for completeness.`

[the remark this answers is not in the document]
Comment (HardeepAsrani) [reply inc/css/class-block-frontend.php:367]: It's fine, else you can see what it is doing and build your own function as per our needs. Also there's `get_block_template` too https://github.com/WordPress/wordpress-develop/blob/5.9/src/wp-includes/block-template-utils.php#L750

[the remark this answers is not in the document]
Comment (Soare-Robert-Daniel) [reply inc/css/class-block-frontend.php:367]: ok. then I will make a function based on `_build_template_result_from_post`

[the remark this answers is not in the document]
Comment (HardeepAsrani) [reply inc/css/class-block-frontend.php:367]: Sure, but make sure to take care of the posts PR as they're more important for 2.0

File: src/blocks/blocks/progress-bar/edit.js [hunk 2]
```diff
@@ -79,6 +82,20 @@ const ProgressBar = ({
...
+			'--percentageColorOuter': attributes.percentageColor,
+			'--percentageColorTooltip': attributes.percentageColor,
+			'--percentageColorAppend': attributes.percentageColor,
+			'--backgroundColor': attributes.backgroundColor,
+			'--borderRadius': ( undefined !== attributes.borderRadius ? attributes.borderRadius : 5 ) + 'px',
+			'--height': ( undefined !== attributes.height ? attributes.height : 30 ) + 'px',
+			'--barBackground': attributes.barBackgroundColor
+		});
+	}, [ blockRef.current, attributes ]);
+
```
Comment (Soare-Robert-Daniel) [inline src/blocks/blocks/progress-bar/edit.js:98]: @HardeepAsrani, using JS to set the CSS vars is the only reliable way to make the CSS work in Editor. We need to convert all the Emotion CSS to this method in this situation.

Comment (Soare-Robert-Daniel) [reply src/blocks/blocks/progress-bar/edit.js:98]: If you referring to `css` var, I can not use it just to render the CSS and inject inline.

Comment (Soare-Robert-Daniel) [reply src/blocks/blocks/progress-bar/edit.js:98]: Also, we do not only need to `only load for FSE`. This can be default method for add CSS

Comment (HardeepAsrani) [reply src/blocks/blocks/progress-bar/edit.js:98]: You can keep it like this working for one block. I'll take a look and let you if I've any input.

Comment (Soare-Robert-Daniel) [reply src/blocks/blocks/progress-bar/edit.js:98]: What do you mean by working for one block?

**Comment by Soare-Robert-Daniel** [comment]
It seams that [get_block_template](https://developer.wordpress.org/reference/functions/get_block_template/) has been changed to `gutenberg_get_block_templates` in the code base: https://cs.github.com/WordPress/gutenberg?q=get_block_templates

But this new function is not available in 5.9.3

**Comment by selul** [comment]
> It seams that [get_block_template](https://developer.wordpress.org/reference/functions/get_block_template/) has been changed to `gutenberg_get_block_templates` in the code base: https://cs.github.com/WordPress/gutenberg?q=get_block_templates
> 
> But this new function is not available in 5.9.3

@Soare-Robert-Daniel I might be wrong, but I know that Gutenberg prefixed functions are available only in the Gutenberg plugin, the prefix is being stripped when the code is merged into core.

**Comment by Soare-Robert-Daniel** [comment]
@selul, you are right; I searched in the Gutenberg repo instead of the core. Thanks, it eliminated some confusion on why I was able to explore it some time ago.

**Comment by Soare-Robert-Daniel** [comment]
@HardeepAsrani, I made a PHP function that can parse the current templates from the post. But we have some tragic situations:

The CSS files are not loading in the posts/pages that call this function. Thus the parsed CSS from the block is useless.
- The pages that belong to the FSE Editor load the CSS files, but they do not call the parsing function.

So we need to find a solution to enqueue our CSS files when a post uses an FSE theme and to detect the FSE only pages so that we can parse them.

**Comment by Soare-Robert-Daniel** [comment]
@HardeepAsrani, made a functional CSS rendering for FSE 🚀 

You can review to see if the approach is correct. Might separate the FSE in a different loader instead of injecting it into the current one as I did.

Also, we need to decide if we want to change the `emotion` CSS library with that function that I made in https://github.com/Codeinwp/otter-blocks/pull/748/commits/0a0fbc6ccb49a9efcd5532a7046a2ac84dc7c5a1

**Comment by HardeepAsrani** [comment]
@Soare-Robert-Daniel You mean to replace emotion everywhere?

**Comment by Soare-Robert-Daniel** [comment]
Something like that

**Comment by HardeepAsrani** [comment]
@Soare-Robert-Daniel If it works nice and covers every situation where we use Emotion then it's fine, also make sure it doesn't trigger CSS again and again adding one thing multiple times. We also target children element from parent using emotion so might be good to check if that works.

File: src/blocks/blocks/countdown/edit.js [hunk 3]
```diff
@@ -176,23 +178,29 @@ const Edit = ({
...
+	useEffect( ()=>{
+		setCSS([
+			`.otter-countdown__display-area .otter-countdown__value {
+				color: ${ attributes.valueColor };
+			}`,
+			`.otter-countdown__display-area .otter-countdown__label {
+				color: ${ attributes.labelColor };
+			}`
+		]);
+	}, [ attributes.valueColor, attributes.labelColor ]);
```
Comment (Soare-Robert-Daniel) [inline src/blocks/blocks/countdown/edit.js:197]: @HardeepAsrani, I made a hook that creates a style node like Emotion `useCSSNode`. We can use it to add CCS code that needs Pseudo-classes, like `hover'. We can also add var, but it is preferable to do it via utility function (made a simple function `cleanCCS`) and to add it inline -- is more reliable this way.

File: src/blocks/blocks/progress-bar/edit.js [hunk 4]
```diff
@@ -91,7 +94,17 @@ const ProgressBar = ({
...
+		'--titleColor': attributes.titleColor,
+		'--percentageColor': attributes.percentageColor,
+		'--percentageColorOuter': attributes.percentageColor,
+		'--percentageColorTooltip': attributes.percentageColor,
+		'--percentageColorAppend': attributes.percentageColor,
+		'--backgroundColor': attributes.backgroundColor,
+		'--borderRadius': attributes.borderRadius !== undefined && ( attributes.borderRadius + 'px' ),
+		'--height': attributes.height !== undefined && ( attributes.height + 'px' ),
+		'--barBackground': attributes.percentageColor
+	};
```
Comment (Soare-Robert-Daniel) [inline src/blocks/blocks/progress-bar/edit.js:107]: @HardeepAsrani, I think this is the most reliable way of handling the CSS var in js. Now, we need to find a 
 reliable solution for handling things like `:hover`.

File: src/blocks/blocks/form/edit.js [hunk 5]
```diff
@@ -690,37 +691,40 @@ const Edit = ({
...
+				${ attributes.submitBackgroundColorHover && `background-color: ${attributes.submitBackgroundColorHover}` }
+			}`
+		]);
+	}, [ attributes.submitBackgroundColor, attributes.submitBackgroundColorHover ]);
+
+	const blockProps = useBlockProps({
+		id: attributes.id,
+		style: inlineStyles,
+		className: cssNodeName
+	});
```
Comment (Soare-Robert-Daniel) [inline src/blocks/blocks/form/edit.js:727]: @HardeepAsrani, this how the CSS will look with the new system

**Comment by HardeepAsrani** [comment]
@Soare-Robert-Daniel I only tested it with Progress Bar and the thing seems fine to me. I saw that you had not migrated one CSS value, so when we do the entire revamp it will be worth making sure we don't forget about anything.

Also, why is there a `cleanCSS` function?

Two things:

- Are there any issues left so far that you're aware of or it's good to go?
- If there are not any that you're aware then we can maybe put this for QA to make sure it is properly and then we can put this in all the blocks.

**Comment by Soare-Robert-Daniel** [comment]
> Also, why is there a `cleanCSS` or `setCSSVars` function or usage of `blockRef`?

There were past methods. It will be cleaned-up

> Maybe we can convert the useEffect part of the code inside the hook:
> 
> ```
> const [ cssNodeName, setCSS ] = useCSSNode();
> 
> useEffect( ()=>{
> 	setCSS([
> 		`.otter-countdown__display-area .otter-countdown__value {
> 			color: ${ attributes.valueColor };
> 		}`,
> 		`.otter-countdown__display-area .otter-countdown__label {
> 			color: ${ attributes.labelColor };
> 		}`
> 	]);
> }, [ attributes.valueColor, attributes.labelColor ]);
> ```
> 
> Or do you think it's better the way it is?

This is the easiest method (I think). With full control of the rendering, same as the blockInit which is also in a `useEffect` . If you have some new ideas, feel free to try them.

> Two things:
> 
> * Are there any issues left so far that you're aware of or it's good to go?

I am trying to discover them. Until now I have had good results. Even in Post with Preview like Table or Mobile which also use an iFrame that disrupts Emotion.

> * If there are not any that you're aware then we can maybe put this for QA to make sure it is properly and then we can put this in all the blocks.

I will convert more blocks with the new method to test them out.

**Comment by Soare-Robert-Daniel** [comment]
After looking at some blocks, most of them did not need emotion. And the most complex CSS was in the Flip block, which we will use for more complex testing.

**Comment by Soare-Robert-Daniel** [comment]
@HardeepAsrani @irinelenache you can try to break the Flip block from this PR.

Check if it is working with:
 - Post / Page with Mobile and Tablet Preview
 - FSE with Mobile and Table Preview
 - Template Parts like Footer or Header
 - Reusable Blocks 

If you succeed, send the instance.

**Comment by Soare-Robert-Daniel** [comment]
I migrate all the blocks to the new system. Most of them did not need Emotion since they only used CSS vars.

Also, I added the case for handling `@media` tags ( needed for Posts blocks). The internal data changer is now exposed if you want to use another identifier or CSS node.

**Comment by HardeepAsrani** [comment]
@Soare-Robert-Daniel You can also make sure emotion isn't being used anywhere & remove it as a dependency.

**Comment by Soare-Robert-Daniel** [comment]
@irinelenache, this is gonna be like testing for a release since all the blocks are affected.

File: inc/class-registration.php [hunk 6]
```diff
@@ -546,8 +546,26 @@ function( $block ) {
 	 * @access  public
 	 */
 	public function enqueue_block_styles( $post ) {
+		$template_parts_content = '';
+
+		if ( function_exists( 'get_block_templates' ) ) {
+			$templates      = get_block_templates( array( 'wp_id' => get_the_ID() ) );
```
Comment (HardeepAsrani) [inline inc/class-registration.php:552]: @Soare-Robert-Daniel This isn't being anywhere in the function.

**Comment by irinelenache** [comment]
@Soare-Robert-Daniel @HardeepAsrani  Tested and here are my findings:

- [ ] The preview issue is still happening for the **advanced heading** https://vertis.d.pr/hPf5Zi , **font awesome icons** in the icon list block https://vertis.d.pr/hufEnN , 
- [ ] The **icon block** doesn't work anymore and it gives an error in customizer https://vertis.d.pr/lwLJyv (Same for the product review)
- [ ] The **leaflet map** breaks in mobile/tablet preview mode https://vertis.d.pr/y5nPGU
- [ ] The **lottie animation** block breaks in mobile/tablet preview mode and gives an error in console https://vertis.d.pr/h6iUFn
- [ ] The **slider** breaks in mobile/tablet preview mode https://vertis.d.pr/6TfuRd
- [ ] The **tabs block** breaks both in mobile/tablet preview and in mobile/tablet frontend
preview: https://vertis.d.pr/fJy4Ay
frontend: https://vertis.d.pr/LVRo7A

^ The mentioned issues happen in FSE mode also.

- [ ] In FSE, **Import blocks from JSON** is missing

**Comment by Soare-Robert-Daniel** [comment]
@irinelenache, any problem with CSS (color, size, etc) so far?

**Comment by irinelenache** [comment]
@Soare-Robert-Daniel No, the only issues i found are the ones i noted

**Comment by Soare-Robert-Daniel** [comment]
- [ ] The **lottie animation** block breaks in mobile/tablet preview mode and gives an error in console https://vertis.d.pr/h6iUFn
- [ ] The **slider** breaks in mobile/tablet preview mode https://vertis.d.pr/6TfuRd

@HardeepAsrani, can you take a look at this. The error seems to happen because when transitioning Mobile/Tablet, the post is moved to an iframe that breaks the references from players (lottie, glide).

Have you encountered something similar?

**Comment by HardeepAsrani** [comment]
@Soare-Robert-Daniel Is that all? Is this also happening to other JS blocks like Maps?

**Comment by Soare-Robert-Daniel** [comment]
> @Soare-Robert-Daniel Is that all? Is this also happening to other JS blocks like Maps?

Those 2 (Lottie, slider) give errors. Leaflet and Tabs are just broken without giving an error on what might happen.

I think WordPress, instead of recreating the block, they are just moving them, thus losing some data on the way (Lottie also loses its image).

**Comment by Soare-Robert-Daniel** [comment]
@irinelenache, I will move the pointed issues to a new Issue since they are unrelated.

You can do a final review on this to check if the CSS is alright. 

Happy bug hunting 🐛

**Comment by Soare-Robert-Daniel** [comment]
The issues related to the advanced header will be also solved separately. They might be related to CSS but are not influenced by this PR.

**Comment by Soare-Robert-Daniel** [comment]
@HardeepAsrani, can you test this PR by running it in your local instance.

Every zip that I made and exported to test on another instance is not fully working. Can you confirm that this is working?

In the other instances, the result of `get_block_templates` is sometimes empty, thus not getting the blocks. 

Test: go to the Home page in the FSE editor, add some blocks in the page and footer or header, and check if CSS is working.

**Comment by Soare-Robert-Daniel** [comment]
@irinelenache, when testing, use the Production ZIP file.

**Comment by Soare-Robert-Daniel** [comment]
@irinelenache @HardeepAsrani adding a block in the header or footer has random behavior, it might work or not in pages that are not Full FSE.

 ⚠️  And by working I mean the loading of the JS file for interactivity. The dynamic CSS is detected and working fine.

The home page which is Full FSE, Circle Counter is working because there is another one in the content that triggers the script loading.
image

Simple post page **_without_** Circle Counter in the content.
image

Simple post page **_with_** Circle Counter in the content.
image

**Comment by Soare-Robert-Daniel** [comment]
I discussed this with @Skystream96, and the problem is persisting even in Stackable.

The general slowness is similar to Otter.  The difference is more sizable on fast typing. On stackable, a very fast written sentence will appear in two half chunks, while in Otter it will appear all at once.

**Comment by Soare-Robert-Daniel** [comment]
@HardeepAsrani, when you have some time, can you check this PR. 

When testing the Home Page for FSE, the content (in our pipeline) comes from the post with id 1, which is the first page. Thus not getting the actual content.

https://github.com/Codeinwp/otter-blocks/blob/2b2c37e7332ebd474a01cf5aa0297f3d322f060d/inc/class-registration.php#L335

And because the ID is wrong (even with `get_the_ID`), the template retrieving is also wrong.

https://github.com/Codeinwp/otter-blocks/blob/2b2c37e7332ebd474a01cf5aa0297f3d322f060d/inc/class-registration.php#L339-L344

Have you encountered this problem?

**Comment by HardeepAsrani** [comment]
@Soare-Robert-Daniel Will be good to know how you've setup your frontpage in FSE so I can have the same setup.

**Comment by Soare-Robert-Daniel** [comment]
> @Soare-Robert-Daniel Will be good to know how you've setup your frontpage in FSE so I can have the same setup.

Nothing fancy, just go to the Home page, insert some blocks (try leaflet, accordion, tabs) and see if it's working.

**Comment by Soare-Robert-Daniel** [comment]
When it comes to templates, the parts of the current template that are going to render are saved in a global variable named ` $_wp_current_template_content`.

All we need is to get the content from the `template-parts` and send it to our pipeline for processing.

https://github.com/Codeinwp/otter-blocks/blob/2a0034165a85aa3b9cb63cfb02af2fed2dcab19e/inc/css/class-block-frontend.php#L435-L456

@HardeepAsrani, one problem remains. We need to see how to determine if the current page is a complete template. Otherwise, the following function will get and parse the content of another post/page like mentioned previously.

https://github.com/Codeinwp/otter-blocks/blob/2b2c37e7332ebd474a01cf5aa0297f3d322f060d/inc/class-registration.php#L335

**Comment by Soare-Robert-Daniel** [comment]
@irinelenache, we know about the issues. We decided to solve them separately in another PR. For the rest of the blocks, the CSS was fine?

**Comment by irinelenache** [comment]
@Soare-Robert-Daniel Yes, it was ok for the other blocks

**Pull request merged**